frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,828 @@
+	Weight, WeightInfo, QueuedSolution, SignedSubmissionsMap, SignedSubmissionIndices,
+	SignedSubmissionNextIndex,
+};
+use codec::{Encode, Decode, HasCompact};
+use frame_support::{
+	storage::bounded_btree_map::BoundedBTreeMap,
+	traits::{Currency, Get, OnUnbalanced, ReservableCurrency},
+	DebugNoBound,
+};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_npos_elections::{is_score_better, CompactSolution, ElectionScore};
+use sp_runtime::{
+	RuntimeDebug,
+	traits::{Saturating, Zero},
+};
+use sp_std::{
+	cmp::Ordering,
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	ops::Deref,
+};
+
+/// A raw, unchecked signed submission.
+///
+/// This is just a wrapper around [`RawSolution`] and some additional info.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
+	/// Who submitted this solution.
+	pub who: AccountId,
+	/// The deposit reserved for storing this solution.
+	pub deposit: Balance,
+	/// The reward that should be given to this solution, if chosen the as the final one.
+	pub reward: Balance,
+	/// The raw solution itself.
+	pub solution: RawSolution<CompactSolution>,
+}
+
+impl<AccountId, Balance, CompactSolution> Ord
+	for SignedSubmission<AccountId, Balance, CompactSolution>
+where
+	AccountId: Ord,
+	Balance: Ord + HasCompact,
+	CompactSolution: Ord,
+	RawSolution<CompactSolution>: Ord,
+{
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.solution
+			.score
+			.cmp(&other.solution.score)
+			.then_with(|| self.solution.cmp(&other.solution))
+			.then_with(|| self.deposit.cmp(&other.deposit))
+			.then_with(|| self.reward.cmp(&other.reward))
+			.then_with(|| self.who.cmp(&other.who))
+	}
+}
+
+impl<AccountId, Balance, CompactSolution> PartialOrd
+	for SignedSubmission<AccountId, Balance, CompactSolution>
+where
+	AccountId: Ord,
+	Balance: Ord + HasCompact,
+	CompactSolution: Ord,
+	RawSolution<CompactSolution>: Ord,
+{
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+pub type SignedSubmissionOf<T> =
+	SignedSubmission<<T as frame_system::Config>::AccountId, BalanceOf<T>, CompactOf<T>>;
+
+pub type SubmissionIndicesOf<T> =
+	BoundedBTreeMap<ElectionScore, u32, <T as Config>::SignedMaxSubmissions>;
+
+/// Mask type which pretends to be a set of `SignedSubmissionOf<T>`, while in fact delegating to the
+/// actual implementations in `SignedSubmissionIndices<T>`, `SignedSubmissionsMap<T>`, and
+/// `SignedSubmissionNextIndex<T>`.
+#[cfg_attr(feature = "std", derive(DebugNoBound))]
```

Comment:
you can also `RuntimeDebugNoBound` to have a stripped implementation in not std

Reply:
Is it useful to have a stripped `Debug` impl in the runtime? I don't know what the tradeoffs are vs this approach.

Reply:
I think for structs that are (or _can_ be) big the debug impl can bloat the code?

Reply:
In that case, this implementation ensures that the runtime stays lean; it only impls `Debug`on `feature = "std"`.

Reply:
there is no necessity, just that I have seen people using RuntimeDebug and RuntimeDebugNoBound in the runtime in order to have stripped implementation of Debug while keeping a standard implementation on std. And thus preventing having to feature gate some stuff.
But your approach here is completely fine.